tensorly/decomposition/_tr.py

```diff
@@ -0,0 +1,102 @@
+import tensorly as tl
+from ..tr_tensor import validate_tr_rank, TRTensor
+
+
+def tensor_ring(input_tensor, rank, mode=0, verbose=False):
+    """Tensor Ring decomposition via recursive SVD
+
+        Decomposes `input_tensor` into a sequence of order-3 tensors (factors) [1]_.
+
+    Parameters
+    ----------
+    input_tensor : tensorly.tensor
+    rank : {int, int list}
+            maximum allowable TR rank of the factors
+            if int, then this is the same for all the factors
+            if int list, then rank[k] is the rank of the kth factor
+    mode : int, default is 0
+            index of the first factor to compute
+    verbose : boolean, optional
+            level of verbosity
+
+    Returns
+    -------
+    factors : TR factors
+              order-3 tensors of the TR decomposition
+
+    References
+    ----------
+    .. [1] Qibin Zhao et al. "Tensor Ring Decomposition" arXiv preprint arXiv:1606.05535, (2016).
+    """
+    rank = validate_tr_rank(tl.shape(input_tensor), rank=rank)
+    n_dim = len(input_tensor.shape)
+
+    # Change order
+    if mode:
+        order = tuple(range(mode, n_dim)) + tuple(range(mode))
+        input_tensor = tl.transpose(input_tensor, order)
+        rank = rank[mode:] + rank[:mode]
+
+    tensor_size = input_tensor.shape
+
+    factors = [None] * n_dim
+
+    # Getting the first factor
+    unfolding = tl.reshape(input_tensor, (tensor_size[0], -1))
+
+    n_row, n_column = unfolding.shape
+    # TODO: Proper warning message
+    if rank[0] * rank[1] > min(n_row, n_column):
+        raise ValueError(f'rank[{mode}] * rank[{mode + 1}] = {rank[0] * rank[1]} is larger than'
+                         'first matricization dimensions.\n'
```

Comment:
Would be good to have a clearer message, e.g.
The mismatch between rank[mode] and rank[0],

The best is to have a message that points precisely the error to the users and helps to fix it. 
We could for instance say the first unfolding is of size (x, y) so will have rank at most min(x, y) and yet got rank[0]*rank[1]=z > min(x, y).